lib/SILOptimizer/Mandatory/TFPartition.cpp

```diff
@@ -2704,10 +2743,149 @@ SILFunction *PartitionCloner::lookupSendReceiveFunction(StringRef fnName,
+      lookupSendReceiveFunction("sendToAccelerator", intFieldSILType, hostLoc);
+  assert(sendFn && "At this point of compilation, the value must be send'able "
+                   "from Swift to TensorFlow.");
+
+  // Create the host->accelerator sends in the host code.
+  for (unsigned i = 0, e = inst->getNumCases(); i != e; ++i) {
+    SILBasicBlock *caseBB = inst->getCase(i).second;
+    SILBuilder BH(caseBB); // Builder for host.
+
+    // The send is at the beginning of each caseBB.
+    BH.setInsertionPoint(&caseBB->front());
+    auto loc = caseBB->front().getLoc();
+    // `caseId` must be of a type conforming to `AccelerableByTensorFlow`, so we
+    // chose Int64 here.
+    auto caseId = createSomeIntegerValue(i, BH, loc, intDecl);
+
+    createHostSend(BH, loc, caseId, tensorComputation, nextSendID,
+                   tensorFlowModule, FP.hostFn.getModule(),
+                   createScalarTensorFn, sendFn);
+  }
+
+  // In accelerator, first recv the caseId from host, and then dispatch to the
+  // right BB. An example SIL snippet is:
+  //
+  // %caseId = builtin "__tfop_tfc.RecvFromHost,tensorId...
+  // %zeroScalarTensor = builtin "__tfop_Const,...
+  // Both operands have type $TensorHandle<Builtin.Int64>
+  // %eq = builtin "__tfop_Equal,$in,$in(%caseId, %zeroScalarTensor)
+  // %cond = builtin "tf_tensor_to_i1"(%eq : $TensorHandle<Builtin.Int1>)
+  // cond_br %cond, bb1, bb2
+  auto BA = getBuilder(); // Builder for accelerator.
+  auto loc = remapLocation(getUserSourceLocation(inst->getDebugLocation()));
+  auto acceleratorCaseId =
+      createAcceleratorReceive(BA, loc, remapType(intFieldSILType),
+                               /*isScalar*/ true, nextSendID, FP.configuration);
+
+  assert(inst->getNumCases() == 2 &&
+         "switch enums with 1 or >2 cases will be supported later.");
+  // Create a scalar const tensor 0, to compare with case id.
+  BuiltinInst *zeroConstTensor = nullptr;
+  {
+    auto *caseZero = BA.createIntegerLiteral(loc, intFieldSILType, 0);
+    SmallVector<SILValue, 4> operands;
+    std::string constOpName = "__tfop_Const";
+    createConstTensorAttrsOnAccel(caseZero, loc, ctx, BA, constOpName,
+                                  operands);
+    FP.configuration.handleDevicePlacement(
+        "Const", /*opDevice*/ getDeviceString(DeviceType::ALL), BA, loc,
+        operands, constOpName);
+    auto subMap = getSingleSubstitutionMapForElementType(
+        intFieldSILType.getASTType(), ctx);
+    auto tensorHandleInt64Ty =
+        convertElementTypeToTensorValueType(intFieldSILType);
+    zeroConstTensor = BA.createBuiltin(loc, ctx.getIdentifier(constOpName),
+                                       tensorHandleInt64Ty, subMap, operands);
+  }
+
+  // Cast to a TensorHandle<Builtin.Int1> value
+  BuiltinInst *condBrOperand = nullptr;
+  {
+    // Omit the metatype attr T for simplicity, and TF graphDef compiler can
+    // infer the type.
+    std::string equalOpName = "__tfop_Equal,$in,$in";
```

Comment:
This would be so much nicer with GraphOp :-)

Reply:
Agree. I look forward to us switching fully over to GraphOp. :)